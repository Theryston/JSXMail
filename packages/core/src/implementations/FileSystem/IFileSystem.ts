export interface IFileSystem {
  getAllDirectoryTree(
    basePath: string,
    options?: IFileSystemOptions
  ): Promise<IDirectoryTree[]>;
  writeFile(path: string, content: string): Promise<void>;
  readUtf8File(path: string): Promise<string>;
  exists(path: string): Promise<boolean>;
  rm(path: string): Promise<void>;
  importFile(path: string): Promise<any>;
}

export interface IFileSystemOptions {
  justFilenameInFilePath?: boolean;
  fileExtensions?: string[];
}

export interface IDirectoryTree {
  path: string;
  children: IDirectoryTree[];
  type: 'file' | 'directory';
}
