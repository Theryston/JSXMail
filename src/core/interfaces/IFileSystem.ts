export interface IDirectoryTree {
  absolutePath: string;
  type: 'file' | 'directory';
  children: IDirectoryTree[];
}

export interface IFileSystem {
  getDirectoryTree(path: string): Promise<IDirectoryTree[]>;
}
