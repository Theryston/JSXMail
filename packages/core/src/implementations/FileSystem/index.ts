import fs from 'fs';
import { IDirectoryTree, IFileSystem, IFileSystemOptions } from './IFileSystem';
import path from 'path';

export class FileSystem implements IFileSystem {
  async getAllDirectoryTree(
    basePath: string,
    options?: IFileSystemOptions
  ): Promise<IDirectoryTree[]> {
    const { justFilenameInFilePath = false, fileExtensions = [] } =
      options || {};

    const entities = await fs.promises.readdir(basePath, {
      withFileTypes: true,
    });
    const directoryTree: IDirectoryTree[] = [];
    for (const entity of entities) {
      const entityPath = path.join(basePath, entity.name);
      if (entity.isDirectory()) {
        const children = await this.getAllDirectoryTree(entityPath, options);
        directoryTree.push({
          children,
          path: entityPath,
          type: 'directory',
        });
      } else {
        if (justFilenameInFilePath) {
          if (
            fileExtensions.length === 0 ||
            fileExtensions.includes(entityPath.split('.').pop())
          ) {
            directoryTree.push({
              children: [],
              path: path.basename(entityPath),
              type: 'file',
            });
          }
        } else {
          if (
            fileExtensions.length === 0 ||
            fileExtensions.includes(entityPath.split('.').pop())
          ) {
            directoryTree.push({
              children: [],
              path: entityPath,
              type: 'file',
            });
          }
        }
      }
    }

    return directoryTree;
  }

  async writeFile(inputPath: string, content: string): Promise<void> {
    const directory = path.dirname(inputPath);
    if (!fs.existsSync(directory)) {
      await fs.promises.mkdir(directory, { recursive: true });
    }
    await fs.promises.writeFile(inputPath, content, 'utf8');
  }

  async readUtf8File(inputPath: string): Promise<string> {
    return await fs.promises.readFile(inputPath, 'utf8');
  }

  async exists(inputPath: string): Promise<boolean> {
    return fs.existsSync(inputPath);
  }

  async rm(inputPath: string): Promise<void> {
    // eslint-disable-next-line
    await (fs.promises as any).rm(inputPath, { recursive: true, force: true });
  }

  // eslint-disable-next-line
  async importFile(inputPath: string): Promise<any> {
    return await import(inputPath);
  }
}
