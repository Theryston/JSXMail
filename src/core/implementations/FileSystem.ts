import fs from 'fs';

import { IFileSystem } from '../interfaces/IFileSystem';

export class FileSystem implements IFileSystem {}
