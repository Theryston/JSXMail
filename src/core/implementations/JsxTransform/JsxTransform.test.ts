import { JsxTransform } from './index';
import { IDirectoryTree, IFileSystem } from '../FileSystem/IFileSystem';
import path from 'path';

let fileSystem: IFileSystem;

const jsxMock = `
import * as React from 'react';
import styled from 'styled-components';

const Container = styled.div\`
background-color: red;
\`;

export function testComponent() {
  return <Container>test</Container>;
}
`;

describe('JsxTransform', () => {
  beforeEach(() => {
    fileSystem = {
      importFile: jest.fn(),
      rm: jest.fn(),
      exists: jest.fn(() => Promise.resolve(true)),
      getAllDirectoryTree: jest.fn(
        async (path: string): Promise<IDirectoryTree[]> => {
          if (path === '') {
            return [
              {
                children: [
                  {
                    children: [],
                    path: 'main.ts',
                    type: 'file',
                  },
                ],
                path: 'src',
                type: 'directory',
              },
              {
                children: [],
                path: 'index.tsx',
                type: 'file',
              },
            ];
          }

          return [
            {
              children: [],
              path: 'main.ts',
              type: 'file',
            },
          ];
        },
      ),
      writeFile: jest.fn(
        async (_path: string, _content: string): Promise<void> => {
          return;
        },
      ),
      readUtf8File: jest.fn(async (_path: string): Promise<string> => {
        return jsxMock;
      }),
    };
  });

  describe('jsxTransform.directory', () => {
    it('should call getAllDirectoryTree with correct path', async () => {
      const jsxTransform = new JsxTransform(fileSystem);
      await jsxTransform.directory('', '');
      expect(fileSystem.getAllDirectoryTree).toHaveBeenCalledWith('', {
        justFilenameInFilePath: true,
      });
      expect(fileSystem.getAllDirectoryTree).toHaveBeenCalledWith('main.ts', {
        justFilenameInFilePath: true,
      });
    });

    it('should call exists with the correct path', async () => {
      const jsxTransform = new JsxTransform(fileSystem);
      await jsxTransform.directory('app', 'dist');
      expect(fileSystem.exists).toHaveBeenCalledWith(path.join('app', 'dist'));
    });

    it('should call rm with the correct path', async () => {
      const jsxTransform = new JsxTransform(fileSystem);
      await jsxTransform.directory('app', 'dist');
      expect(fileSystem.rm).toHaveBeenCalledWith(path.join('app', 'dist'));
    });
  });

  describe('jsxTransform.file', () => {
    it('should call readUtf8File with correct path', async () => {
      const jsxTransform = new JsxTransform(fileSystem);
      await jsxTransform.file('main.tsx', '');
      expect(fileSystem.readUtf8File).toHaveBeenCalledWith('main.tsx');
    });

    it('should call writeFile with correct path', async () => {
      const jsxTransform = new JsxTransform(fileSystem);
      await jsxTransform.file('main.js', 'dist');
      expect(fileSystem.writeFile).toHaveBeenCalledWith(
        path.join('dist', 'main.js'),
        expect.any(String),
      );
    });
  });

  describe('jsxTransform.code', () => {
    it('should transform code', async () => {
      const jsxTransform = new JsxTransform(fileSystem);
      const compiledCode = await jsxTransform.code(jsxMock);
      expect(compiledCode).toMatchSnapshot();
    });

    it('should transform code with custom runtime', async () => {
      const jsxTransform = new JsxTransform(fileSystem);
      const compiledCode = await jsxTransform.code(jsxMock, {
        customRuntime: 'customRuntime',
      });
      expect(compiledCode).toMatchSnapshot();
    });
  });
});
