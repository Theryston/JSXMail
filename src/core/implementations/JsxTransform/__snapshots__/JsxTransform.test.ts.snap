// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsxTransform jsxTransform.code should transform code 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.testComponent = testComponent;

var React = _interopRequireWildcard(require(\\"react\\"));

var _styledComponents = _interopRequireDefault(require(\\"styled-components\\"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== \\"function\\") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== \\"object\\" && typeof obj !== \\"function\\") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== \\"default\\" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const Container = _styledComponents.default.div.withConfig({
  displayName: \\"file__Container\\"
})([\\"background-color:red;\\"]);

function testComponent() {
  return /*#__PURE__*/React.createElement(Container, null, \\"test\\");
}"
`;

exports[`JsxTransform jsxTransform.code should transform code with custom runtime 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.testComponent = testComponent;

var React = _interopRequireWildcard(require(\\"react\\"));

var _styledComponents = _interopRequireDefault(require(\\"styled-components\\"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== \\"function\\") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== \\"object\\" && typeof obj !== \\"function\\") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== \\"default\\" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/** @jsx customRuntime */
const Container = _styledComponents.default.div.withConfig({
  displayName: \\"file__Container\\"
})([\\"background-color:red;\\"]);

function testComponent() {
  return customRuntime(Container, null, \\"test\\");
}"
`;
